### PR TITLE
rework android permissions

### PR DESCRIPTION
**Pull request by sleushunou**

### CODE REVIEW

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs [hunk 1]
```diff
@@ -75,7 +75,9 @@ private async Task<PermissionStatus> CommonCheckWithRequestAsync<T>()
                 return permissionStatus;
             }
 
-            if (permissionStatus == PermissionStatus.Denied && IsPermissionRequested<T>())
+            if (permissionStatus == PermissionStatus.Denied
+                && IsPermissionDeniedEver<T>()
+                && !Xamarin.Essentials.Permissions.ShouldShowRationale<T>())
```
Comment (TimofeyBurak) [inline Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs:80]: As I understand, `IPermissionService` is our proxy for Essentials.Permissions. Maybe we should move this method into `PermissionService` and extend interface?

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs [hunk 2]
```diff
@@ -43,19 +43,19 @@ public void SetPermissionDialogService(IPermissionsDialogService permissionsDial
...
         {
-            Preferences.Set(GetPermissionRequestedKey<T>(), value);
+            Preferences.Set(GetPermissionDeniedEverKey<T>(), value);
         }
 
-        private string GetPermissionRequestedKey<T>()
+        private string GetPermissionDeniedEverKey<T>()
             where T : BasePermission
         {
             return $"{nameof(PermissionsManager)}_IsPermissionRequested_{typeof(T).Name}";
```
Comment (TimofeyBurak) [inline Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs:61]: Should we also update key value? Now that the meaning has changed - previous values are not valid either way

**Review by TimofeyBurak — CHANGES_REQUESTED** [review]

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs (hunk 1 above)
Comment (wcoder) [reply Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs:80]: Currently, [XToolkit.Permissions fully depend on Xamarin.Essentials](https://softeq.github.io/XToolkit.WhiteLabel/articles/xtoolkit/permissions.html) for Services and Managers, contains Interfaces over Essentials and some custom behavior only (double-dialogs).

TODO: merge platform-specific services into one
(the idea is: single service for extend XE, and 2 managers for custom UI behavior)

**Review by wcoder — CHANGES_REQUESTED** [review]
Now, the Preferences value has another sense, should be renamed or migrated.

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs (hunk 2 above)
Comment (sleushunou) [reply Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs:61]: May have problems with backward compatibility. If user already press "don's show never again", system permission will never shown, and we can not navigate user to settings

**Comment by wcoder** [comment]
The building issue was in old link to XCode, fixed by #462

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs (hunk 2 above)
Comment (sleushunou) [reply Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs]: Migration added

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs (hunk 1 above)
Comment (sleushunou) [reply Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs]: Moved to IPermissionService

**Comment by wcoder** [comment]
@sleushunou could you please rebase latest master branch?

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs [hunk 3]
```diff
@@ -43,29 +43,51 @@ public void SetPermissionDialogService(IPermissionsDialogService permissionsDial
...
+            var requestedKeyName = GetPermissionRequestedKey<T>();
+            if (Preferences.ContainsKey(requestedKeyName)
+                && Preferences.Get(requestedKeyName, false))
+            {
+                if (permissionStatus == PermissionStatus.Denied)
+                {
+                    SetPermissionDenied<T>(true);
+                }
+
+                Preferences.Remove(GetPermissionRequestedKey<T>());
```
Comment (wcoder) [inline Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs:81]: use `.Remove(requestedKeyName)`

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs [hunk 4]
```diff
@@ -43,29 +43,51 @@ public void SetPermissionDialogService(IPermissionsDialogService permissionsDial
...
         private void OpenSettings()
         {
             _permissionsService.OpenSettings();
         }
 
+        private void ApplyKeysMigration<T>(PermissionStatus permissionStatus)
+            where T : BasePermission, new()
+        {
+            var requestedKeyName = GetPermissionRequestedKey<T>();
+            if (Preferences.ContainsKey(requestedKeyName)
```
Comment (wcoder) [inline Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs:73]: I propose a way like this:

```
if (!containsKey)
	return

if (keyValue = true) {
  // do migration
}

// remove key 
```

**Review by wcoder — CHANGES_REQUESTED** [review]
LGTM, but there are a couple of comments

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs (hunk 3 above)
Comment (sleushunou) [reply Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs]: fixed

File: Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs (hunk 4 above)
Comment (sleushunou) [reply Softeq.XToolkit.Permissions.Droid/PermissionsManager.cs]: +

**Review by wcoder — APPROVED** [review]

**Review by TimofeyBurak — APPROVED** [review]

**Pull request merged**